lib/puppet/network/http/connection.rb

```diff
@@ -213,7 +213,7 @@ def do_request(request, options)
               current_request[header] = value
             end
           when 429, 503
-            connection.finish
+            connection.finish if connection.started?
```

Comment:
I don't think this is right. If the remote side sends `Connection: close`, then [ruby will determine that the socket should not be kept alive, and will close the `@socket`](https://github.com/ruby/ruby/blob/v2_5_3/lib/net/http.rb#L1562-L1568), but it never sets `@started = false`. So the connection will be left in an inconsistent state. It's only when we call `finish` that the [`@started` flag is changed and the `@socket` is set to nil](https://github.com/ruby/ruby/blob/v2_5_3/lib/net/http.rb#L1007-L1016). In that scenario `@socket.close` to be called twice, but I believe the second close it ignored. I'll confirm that.

Reply:
Yeah, the second close is ignored. https://github.com/ruby/ruby/blob/c4b66522238e9e1e164fca2b9497a80acbfc397e/ext/openssl/lib/openssl/ssl.rb#L371